File: backend/app/schemas/device.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class DeviceCreate(BaseModel):
    device_model: str | None = None
    os_version: str | None = None
    app_version: str | None = None
    phone_number: str | None = None
    imei: str | None = None
    assigned_driver_id: uuid.UUID | None = None


class DeviceUpdate(BaseModel):
    device_model: str | None = None
    os_version: str | None = None
    app_version: str | None = None
    phone_number: str | None = None
    assigned_driver_id: uuid.UUID | None = None
    status: str | None = None
    config: dict | None = None


class DeviceResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    device_token: str
    device_model: str | None
    os_version: str | None
    app_version: str | None
    phone_number: str | None
    imei: str | None
    assigned_driver_id: uuid.UUID | None
    status: str
    battery_level: int | None
    last_heartbeat_at: datetime | None
    last_location_lat: float | None
    last_location_lng: float | None
    config: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DeviceCommandCreate(BaseModel):
    device_id: uuid.UUID
    command_type: str
    payload: dict = {}


class DeviceCommandResponse(BaseModel):
    id: uuid.UUID
    device_id: uuid.UUID
    command_type: str
    payload: dict
    status: str
    result: dict | None
    issued_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class BulkCommandRequest(BaseModel):
    device_ids: list[str]
    command_type: str
    payload: dict = {}

File: frontend/src/types/device.ts
export interface Device {
  id: string;
  tenant_id: string;
  device_token: string;
  device_model: string | null;
  os_version: string | null;
  app_version: string | null;
  phone_number: string | null;
  imei: string | null;
  assigned_driver_id: string | null;
  status: string;
  battery_level: number | null;
  last_heartbeat_at: string | null;
  last_location_lat: number | null;
  last_location_lng: number | null;
  config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface DeviceCreate {
  device_model?: string;
  os_version?: string;
  app_version?: string;
  phone_number?: string;
  imei?: string;
  assigned_driver_id?: string;
}

export interface DeviceUpdate {
  device_model?: string;
  os_version?: string;
  app_version?: string;
  phone_number?: string;
  assigned_driver_id?: string;
  status?: string;
  config?: Record<string, unknown>;
}

export interface DeviceFilters {
  status?: string;
  driver_id?: string;
  search?: string;
  page?: number;
  per_page?: number;
}

export interface DeviceCommand {
  id: string;
  device_id: string;
  command_type: string;
  payload: Record<string, unknown>;
  status: string;
  result: Record<string, unknown> | null;
  issued_at: string;
  completed_at: string | null;
}

export interface DeviceCommandCreate {
  device_id: string;
  command_type: string;
  payload?: Record<string, unknown>;
}

export interface BulkCommandRequest {
  device_ids: string[];
  command_type: string;
  payload?: Record<string, unknown>;
}
